lib/std/mem.zig

```diff
@@ -635,13 +635,76 @@ test "lessThan" {
 /// Compares two slices and returns whether they are equal.
 pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
     if (a.len != b.len) return false;
-    if (a.ptr == b.ptr) return true;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
+    if (@sizeOf(T) == 0) return true;
+
+    if (std.meta.hasUniqueRepresentation(T)) return eqlBytes(std.mem.sliceAsBytes(a), std.mem.sliceAsBytes(b));
+
     for (a, b) |a_elem, b_elem| {
         if (a_elem != b_elem) return false;
     }
     return true;
 }
 
+/// std.mem.eql heavily optimized for slices of bytes.
+pub fn eqlBytes(a: []const u8, b: []const u8) bool {
```

Comment:
> The idea here is to split this api into 2 different functions. The normal eql function, which works on anything, and the much more optimized eqlBytes, which specifically optimizes bytes. eql calls eqlBytes if T is u8.

Is there a reason that this is public?

Reply:
It's mainly that I didn't see a reason *not* to make it public :) 

If the user wants to use the eqlBytes function, good on them. It both makes the code easier to read, as the purpose is in the name, and removes an extra parameter that doesn't need to exist. I'm not the typo of person to advocate for squeezing out every character and making the names smaller, but I think it's still easier in the long run.

Reply:
I'm not sure it is necessary to make it public. Per the Zen:

```
 * Only one obvious way to do things.
```

Making it public means that it is objectively fine to write both `std.mem.eql(u8, a, b)` and `std.mem.eqlBytes(a, b)`. I feel like we're better off just allowing one if they do the exact same thing.

Furthermore, people may decide to manually cast their non-`u8` slices to satisfy `eqlBytes`'s `[]const u8` parameter by manually using `std.mem.toBytes`. Therefore, leaving both options to produce the same behavior could lead to unnecessarily uses of `std.mem.eqlBytes`.

Reply:
You bring up a very good point. With this in consideration I agree that it should not be made public, and I'll change that :)

Second point increases my willingness to do it. People might not realize how `sliceAsBytes` behaves with padding, and this'll cause issues I imagine.

Thanks!